### PR TITLE
Refactor Bash Scripts for Enhanced Readability and Robustness

### PR DESCRIPTION
**Pull request by Adriel007**
This pull request encompasses a series of commits aimed at improving the readability, portability, and robustness of various Bash scripts. Here's a summary of the changes:

- Improved Variable Naming: Descriptive variable names were adopted for better understanding.
- OS Detection with `case` Statements: `case` statements were employed for OS detection, enhancing script portability.
- Consistent Output with `printf`: Replaced `echo` with `printf` for consistent output formatting.
- Dependency Checks: Scripts now check for dependencies like `wget` before usage to prevent execution errors.
- Enhanced Error Messages: Error messages were enhanced for better debugging and understanding failure cases.
- Function Organization: Scripts were reorganized into functions for better maintainability.
- Error Handling: Added error handling and informative messages for failure cases, improving script reliability.
- Consistent Indentation and Formatting: Ensured consistent indentation and formatting for readability.
- Variable Quoting: Quoted variables to prevent word splitting and pathname expansion.
- Comments and Documentation: Added comments and updated usage messages for clarity and understanding of the scripts' purpose and individual sections.

These changes aim to provide more robust, maintainable, and understandable solutions for various tasks including Kubeflow installation, KubeRay installation, Minio setup, NGINX installation and management, Kind cluster management, and NGINX Ingress resource management.

### CODE REVIEW

File: transforms/code/malware/install_clamd_mac.sh [hunk 1]
```diff
@@ -1,26 +1,35 @@
...
-sudo chown $(id -u):$(id -g) /var/run/clamav
-
-# Create a direcotry for a database of ClamAV
-if [ ! -e /var/lib/clamav ]; then
-    sudo mkdir -p /var/lib/clamav
-fi
-sudo chown $(id -u):$(id -g) /var/lib/clamav
-
-# Update a database of ClamAV
+# Copy and edit clamd.conf
```
Comment (daw3rd) [inline transforms/code/malware/install_clamd_mac.sh:19]: This all looks very good.  However, when I run this on a Mac M1, it fails `make test-src`.  If i use the original version it does not fail.  I tried adding back the `\cp` in place of `cp`, but no help.
```
make clean venv test-src
...
Warning: clamav 1.3.1 is already installed and up-to-date.
To reinstall 1.3.1, run:
  brew reinstall clamav
ClamAV update process started at Fri May  3 09:24:24 2024
daily.cld database is up-to-date (version: 27264, sigs: 2060601, f-level: 90, builder: raynman)
main.cvd database is up-to-date (version: 62, sigs: 6647427, f-level: 90, builder: sigmgr)
bytecode.cvd database is up-to-date (version: 335, sigs: 86, f-level: 90, builder: raynman)
source venv/bin/activate;       \
	export PYTHONPATH=../src;       \
	cd test; pytest -s .
ERROR: Please define server type (local and/or TCP).
========================================================================= test session starts =========================================================================
platform darwin -- Python 3.10.11, pytest-8.2.0, pluggy-1.5.0
rootdir: /Users/dawood/git/ariel007/data-prep-lab/transforms/code/malware/test
plugins: anyio-4.3.0
collecting ... 09:24:28 INFO - Clamd process is not running. Start clamd process.
ERROR: Please define server type (local and/or TCP).
collected 1 item / 1 error 
```

**Review by daw3rd — CHANGES_REQUESTED** [review]

File: transforms/code/malware/install_clamd_mac.sh (hunk 1 above)
Comment (Adriel007) [reply transforms/code/malware/install_clamd_mac.sh:19]: I will fix this

File: transforms/code/malware/install_clamd_mac.sh [hunk 2]
```diff
@@ -1,26 +1,35 @@
...
-# Copy and edit config files.
-\cp -f $(brew --prefix)/etc/clamav/clamd.conf.sample $(brew --prefix)/etc/clamav/clamd.conf
-sed -i '' -e 's/^Example/# Example/' $(brew --prefix)/etc/clamav/clamd.conf
-echo "DatabaseDirectory /var/lib/clamav" >> $(brew --prefix)/etc/clamav/clamd.conf
-echo "LocalSocket /var/run/clamav/clamd.ctl" >> $(brew --prefix)/etc/clamav/clamd.conf
-\cp -f $(brew --prefix)/etc/clamav/freshclam.conf.sample $(brew --prefix)/etc/clamav/freshclam.conf
-sed -i '' -e 's/^Example/# Example/' $(brew --prefix)/etc/clamav/freshclam.conf
-echo "DatabaseDirectory /var/lib/clamav" >> $(brew --prefix)/etc/clamav/freshclam.conf
-
-# Create a directory for a local unix socket
```
Comment (daw3rd) [inline transforms/code/malware/install_clamd_mac.sh:13]: Can you restore this and the next comment

**Review by daw3rd — CHANGES_REQUESTED** [review]

File: kind/hack/tools/install_kuberay.sh [hunk 3]
```diff
@@ -1,60 +1,67 @@
...
-		header_text "Installing KubeRay"
-		deploy
-		;;
-	*)
-		usage
-		;;
-esac
-
+    cleanup)
+        echo "Uninstalling KubeRay..."
```
Comment (roytman) [inline kind/hack/tools/install_kuberay.sh:52]: Please replace `echo` by  `header_text`

File: kind/hack/tools/install_kuberay.sh [hunk 4]
```diff
@@ -1,60 +1,67 @@
...
+        delete
+        ;;
+    deploy-wait)
+        echo "Deploying and waiting for KubeRay..."
+        deploy
+        wait_for_kuberay
```
Comment (roytman) [inline kind/hack/tools/install_kuberay.sh:58]: the deployment fails here, with the message:

> Error: INSTALLATION FAILED: cannot re-use a name that is still in use
Error: Failed to install kuberay-operator.

in order to reduce installation time, the making process deploys everything and, after that, waits for one by another. 
Please remove L57 (`deploy`) from the deploy-wait operation.

File: kind/hack/tools/install_nginx.sh [hunk 5]
```diff
@@ -1,53 +1,52 @@
...
-		usage
-		;;
+case "$operation" in
+    cleanup)
+        header_text "Uninstalling NGINX"
+        uninstall_nginx
+        ;;
+    deploy-wait)
+        header_text "Deploying and waiting for NGINX"
+        deploy_nginx
```
Comment (roytman) [inline kind/hack/tools/install_nginx.sh:42]: please remove L42

File: kind/hack/tools/install_kuberay.sh [hunk 6]
```diff
@@ -1,60 +1,67 @@
...
-
-	if [[ $EXIT_CODE -ne 0 ]]
-	then
-		echo "KubeRay Deployment unsuccessful. Not all pods running"
-		exit $EXIT_CODE
-	fi
+# Function to wait for KubeRay deployment
+wait_for_kuberay() {
+    echo "Waiting for KubeRay deployment..."
+    wait_for_pods "kuberay" "$MAX_RETRIES" "$SLEEP_TIME" || { echo "Error: KubeRay deployment unsuccessful. Not all pods running."; exit 1; }
```
Comment (roytman) [inline kind/hack/tools/install_kuberay.sh:33]: The command fails here:
./install_kuberay.sh: line 33: wait_for_pods: command not found,

you removed `source ../common.sh`, please return it

File: kind/hack/tools/install_kuberay.sh [hunk 7]
```diff
@@ -1,60 +1,67 @@
...
-source ../common.sh
+# This script manages the deployment of KubeRay components.
 
+# Define global variables
+op="$1"
 SLEEP_TIME="${SLEEP_TIME:-30}"
 MAX_RETRIES="${MAX_RETRIES:-5}"
-EXIT_CODE=0
 
+# Define functions
```
Comment (roytman) [inline kind/hack/tools/install_kuberay.sh:10]: do we need this line? looks like it duplicates L12

File: kind/hack/tools/ingress.sh [hunk 8]
```diff
@@ -1,40 +1,46 @@
...
 
-source ${ROOT_DIR}/hack/common.sh
+readonly ROOT_DIR="/path/to/root/dir"
 
-deploy() {
-	kubectl apply -f ${ROOT_DIR}/hack/ray_api_server_ingress.yaml
-if [[ "${DEPLOY_KUBEFLOW}" -eq 1 ]]; then
-	kubectl apply -f ${ROOT_DIR}/hack/kfp_ingress.yaml
-fi
+source "${ROOT_DIR}/hack/common.sh"
```
Comment (roytman) [inline kind/hack/tools/ingress.sh:7]: the path is wrong, 

> ./ingress.sh: line 7: /path/to/root/dir/hack/common.sh: No such file or directory

File: kind/hack/populate_minio.sh [hunk 9]
```diff
@@ -1,20 +1,45 @@
...
+    echo "Copying data for $module"
+    mc cp --recursive "${source_dir}/${module}/test-data/input/" "kfp/test/${dest}/input"
+}
+
+# Set Minio alias
+echo "Creating Minio alias"
+mc alias set "$minio_alias" "$minio_url" "$minio_access_key" "$minio_secret_key"
+
+# Create test buckets
+echo "Creating test buckets"
```
Comment (roytman) [inline kind/hack/populate_minio.sh:25]: we don't need a separate bucket for each tested transformer. The "folder" prefix is enough.

File: kind/hack/populate_minio.sh [hunk 10]
```diff
@@ -1,20 +1,45 @@
...
+minio_url="http://localhost:8080"
+minio_alias="kfp"
+minio_access_key="minio"
+minio_secret_key="minio123"
+source_dir="${ROOT_DIR}/../transforms"
 
+# Function to copy data
+copy_data() {
+    local module="$1"
+    local src="$2"
```
Comment (roytman) [inline kind/hack/populate_minio.sh:13]: this argument is not used

File: kind/hack/tools/ingress.sh [hunk 11]
```diff
@@ -1,40 +1,46 @@
 #!/usr/bin/env bash
 
-op=$1
+set -euo pipefail
 
-source ${ROOT_DIR}/hack/common.sh
+readonly ROOT_DIR="/path/to/root/dir"
```
Comment (roytman) [inline kind/hack/tools/ingress.sh:5]: what is `/path/to/root/dir` ?

> Download helper-functions.sh

**Review by roytman — CHANGES_REQUESTED** [review]
Hi @Adriel007 , welcome, and thank you for your PR.

See several comments.

**Pull request closed without merging**